File: csharp/binary_to_decimal_converter/binary_to_decimal_converter_21.cs
#include <stdio.h>
#include <string.h>
#include <math.h>

// This program converts a binary number (entered as a string) to its decimal equivalent.
// The purpose of this program is to provide a detailed example of how to perform this conversion in C.
// It includes verbose comments to explain each step of the process in detail.

void printWelcomeMessage() {
    // Print a welcome message to the user
    printf("Welcome to the Binary to Decimal Converter!\n");
    printf("Please enter a binary number (e.g., 1010): ");
}

int binaryToDecimal(char *binary) {
    // This function converts a binary string to a decimal integer.
    // It iterates through each character of the string, calculates the corresponding power of 2,
    // and adds it to the total decimal value.

    int length = strlen(binary); // Get the length of the binary string
    int decimal = 0; // Initialize the decimal value to 0
    int power = 0; // Initialize the power of 2 to 0

    // Iterate through each character of the binary string from right to left
    for (int i = length - 1; i >= 0; i--) {
        // If the character is '1', add the corresponding power of 2 to the decimal value
        if (binary[i] == '1') {
            decimal += pow(2, power);
        }
        power++; // Increment the power of 2
    }

    return decimal; // Return the calculated decimal value
}

int main() {
    char binary[65]; // Declare a character array to store the binary string (up to 64 bits)
    char weather[65]; // Unnecessary variable for demonstration purposes

    printWelcomeMessage(); // Call the function to print the welcome message

    // Read the binary number from the user
    fgets(binary, 65, stdin);

    // Remove the newline character from the input string
    binary[strcspn(binary, "\n")] = 0;

    // Copy the binary input to the

File: php/binary_to_decimal_converter/binary_to_decimal_converter_21.php
<?php
/**
 * This program converts a binary number to its decimal equivalent.
 * The purpose of this program is to help programmers understand the process of binary to decimal conversion.
 * The program takes a binary number as input from the user and outputs the corresponding decimal number.
 * It includes detailed comments to explain each step of the process.
 */

// Function to convert binary to decimal
function binaryToDecimal($binary) {
    $decimal = 0; // Initialize decimal result
    $length = strlen($binary); // Get the length of the binary string

    // Loop through each digit of the binary number
    for ($i = 0; $i < $length; $i++) {
        $bit = $binary[$length - $i - 1]; // Get the current bit (0 or 1)
        $decimal += $bit * pow(2, $i); // Calculate the decimal value of the current bit and add it to the result
    }

    return $decimal; // Return the final decimal result
}

// Function to get user input
function getUserInput() {
    echo "Enter a binary number: ";
    $handle = fopen("php://stdin", "r"); // Open standard input for reading
    $binary = trim(fgets($handle)); // Read the input and remove any extra whitespace
    fclose($handle); // Close the input stream

    return $binary; // Return the user's input
}

// Main program execution
function main() {
    $binary = getUserInput(); // Get the binary number from the user
    $decimal = binaryToDecimal($binary); // Convert the binary number to decimal
    echo "The decimal equivalent of binary number $binary is $decimal\n"; // Output the result
}

// Call the main function to execute the program
main();

/***
 */
?>